### PR TITLE
Fix incorrect usage of CUDACachingAllocator

### PR DESCRIPTION
**Pull request by xwang233**
We need an object to hold the ownership of allocated memory in the scope, instead of directly using the raw pointer.

### CODE REVIEW

**Comment by xwang233** [comment]
cc @ptrblck

File: aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cu [hunk 1]
```diff
@@ -58,9 +58,10 @@ static void apply_batched_inverse_lib(Tensor& self, Tensor& self_inv, Tensor& in
       can_start.record(main_stream);
       can_start.block(main_stream);
 
-      int* pivot = reinterpret_cast<int*>(allocator.allocate(sizeof(int) * n).get());
+      auto dataPtr = allocator.allocate(sizeof(int) * n);
       _apply_single_inverse_helper<scalar_t>(
-        &self_data[i * self_mat_stride], &self_inv_data[i * self_inv_mat_stride], pivot, p_infos + i, n);
+        &self_data[i * self_mat_stride], &self_inv_data[i * self_inv_mat_stride],
+        reinterpret_cast<int*>(dataPtr.get()), p_infos + i, n);
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cu:64]: this predates this pr, but, shouldn't need a reinterpret cast here, static cast from void* should be fine

**Review by ezyang — APPROVED** [review]
Thanks. So when we gonna get "disable caching allocator and cuda-memcheck" ci job 👍

**Comment by ezyang** [comment]
need to merge

**Comment by xwang233** [comment]
It conflicted with my own changes. 😂 Resolved.

**Pull request closed without merging**